### PR TITLE
Feature to allow multiple widgets

### PR DESCRIPTION
**Pull request by stevegal**
This is in response to #2481

I'll update this as I can prove out more bits, but I'll need some **help**, especially around testing of the different widget types. 
**Note** this will change your mongo database after use, so make sure you don't try this against your live data.
### TODO:

- [ ] sync dashboard - not sure what this is doing
### Test List:
- [ ] Widget delete on dashboard networking
- [X] Update Config from widget data not global
   - [X] repo
   - [ ] build
   - [ ] chatops
- [ ] test multiple performance widgets
- [ ] test multiple rally widgets
- [ ] chatops widget
- [ ] deploy widget
### Confirmed working:
- [X] test multiple repo widgets
- [X] editDashboard - refactor to not work just of types
- [X] build widgets
- [X] feature widgets
- [X] test multiple code analysis widgets

Comments and feedback appreciated. Thanks already to @dardanbekteshi for help in testing this out so far for me.

Note: needs to build along with changes in https://github.com/Hygieia/hygieia-core/pull/40

### CODE REVIEW

**Comment by stevegal** [comment]
@rvema @matthewbeasley looking for any comments on this please!

**Comment by stevegal** [comment]
any comments yet?

I’m currently working on the edit dashboard... Is there a reason why this is so different from the create?

**Comment by dardanbekteshi** [comment]
> any comments yet?
> 
> I’m currently working on the edit dashboard... Is there a reason why this is so different from the create?

@stevegal I will test the new changes this Friday and come back to you with comments/feedback.

**Comment by dardanbekteshi** [comment]
Hi @stevegal, some observations from testing the PR on my local setup:

- Deletion
Deleting a widget doesn't currently work, in particular the below for loop removes all collectorItems of a given collectorType. https://github.com/stevegal/Hygieia/blob/5206b98383a8aa8151e8a9d6d42f8f0cdff59702/api/src/main/java/com/capitalone/dashboard/service/DashboardServiceImpl.java#L577-L579
delete_feature_1
delete_build_3
As you can see, when one deletes a widget of one type, all the other widgets "lose" the data instantly.
Can maybe these changes from my branch be helpful?https://github.com/dardanbekteshi/Hygieia/blob/8f126a8ef8c2f2cab6cfc54346c7ab74f6198cae/api/src/main/java/com/capitalone/dashboard/service/DashboardServiceImpl.java#L540-L550
https://github.com/dardanbekteshi/Hygieia/blob/8f126a8ef8c2f2cab6cfc54346c7ab74f6198cae/api/src/main/java/com/capitalone/dashboard/service/DashboardServiceImpl.java#L569

-  Build widget creation & update
The first widget creation works fine. When trying to create and/or update the 2nd one, the "creation" widget gets prefilled with the options of the 1st build.
build_2_creation
Basically the 1st textbox **BuildJob** always takes the value of the 1st widget's job and you need to manually change it even after an update of an already defined 2nd widget.
**Note**: the two other options, alert criteria and threshold work fine (as expected).

- Repo widget
Similar problem as in Build. When creating the 2nd widget, all the options get prefilled from the 1st widget. 
repo_2_creation

**Comment by stevegal** [comment]
@dardanbekteshi thanks for the detailed report and the code suggestions. My preference (as yours) is to work collaboratively on this... It is a relatively big job for just interested parties to commit to. I know you opened up your fork to PRs first but I had made a bit of progress... I'm willing to work either way on this (me contributing to your PR or vice-versa), I can't wait for this to work as it opens up so suck flexibility to people to use this how they want. Just let me know how you want to proceed (your fork or mine!). 
Thanks again!

**Comment by dardanbekteshi** [comment]
@stevegal It is indeed a relatively big task and I think you made a good point.  
I have no problem on working on either fork, IMO what matters is to join forces, the other stuff is just details. But I'd suggest we work on your branch since that is the one with the highest activity.

**Comment by msoyer78** [comment]
Hi,
  I need the same thing, how is this progressing ?

**Comment by stevegal** [comment]
@msoyer78 - so far so good. Basic functionality is present as I started on create dashboard first, but I'm working on update/delete widgets at the moment. It's slow as I'm working on it in only when I have free time. Anyway... check it out and let me know what you think.

**Comment by stevegal** [comment]
Editing dashboards now possible
screenshot 2019-02-08 at 17 56 04

Repo config now sorted to pull in its config and not master. Delete should also work

**Comment by stevegal** [comment]
@rvema @matthewbeasley I'm not sure how much further I can take this with my limited collector options. You can see what I do with the repo widget, basically query for the collectorItem and fill in from there. What do you think?

**Comment by chtompki** [comment]
This has merge conflicts on the branch as well as compiler errors. We should remediate those before considering pulling this in.

**Comment by stevegal** [comment]
I’ve raised an equivalent pr on the core project as well and pointed that out in the comments above. It requires modifications on both sides

**Comment by shishirchaturvedi** [comment]
@chtompki @stevegal This a great contribution. How soon can we complete/merge this PR?

**Comment by chtompki** [comment]
We’ll want to resolve #2766 (I have an idea here that I’ll ask about tomorrow) before hand. But we agree

**Comment by stevegal** [comment]
I filled in the old contributors agreement if that helps things?

**Comment by poojithkumar** [comment]
@stevegal  is any one collectors support for mutiple widget now ?

**Comment by stevegal** [comment]
I think I’ve done most ui widgets, but can’t be sure they work as I don’t have the ability to test them all out manually

**Comment by poojithkumar** [comment]
change core dependency like this in all pom files
<dependency>
            <groupId>com.capitalone.dashboard</groupId>
            <artifactId>core</artifactId>
            <version>3.1.0-SNAPSHOT</version>
        </dependency>

**Comment by poojithkumar** [comment]
still i am getting error in deletion of widget other widget loose data

**Comment by poojithkumar** [comment]
Capture


widget number counting all widget of same type created in all. it should start from 1 if other is deleted

**Comment by poojithkumar** [comment]
is there any way to change widget name in view file. i want name of the widget as a component name . that's more meaningful name than numbers.


thanks

**Comment by stevegal** [comment]
@poojithkumar yeah... it would be quite easy I think to update the names.But it's already a big PR without adding even more to it.

The dropped data if you delete one widget... does that relate to all widgets or just particular types? I thought I solved it for a few of them, so I'd be interested to know which I've missed (or indeed if any work!).

Thanks

**Comment by poojithkumar** [comment]
If u suggest the method to do so .it will be great help for me. (Plz share
advice)
I tried only for build widget.

Thanks  in advance.

**Comment by poojithkumar** [comment]
i am getting this error while building.
i already build core also
thanks
                 

 [INFO] /workspace/hygieia-poojith/Hygieia/api/src/main/java/com/capitalone/dashboard/auth/ldap/CustomUserDetailsContextMapper.java: Some input files use unchecked or unsafe operations.
[INFO] /workspace/hygieia-poojith/Hygieia/api/src/main/java/com/capitalone/dashboard/auth/ldap/CustomUserDetailsContextMapper.java: Recompile with -Xlint:unchecked for details.
[INFO] Some messages have been simplified; recompile with -Xdiags:verbose to get full output
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] /workspace/hygieia-poojith/Hygieia/api/src/main/java/com/capitalone/dashboard/request/DashboardRequest.java:[141,11] incompatible types: java.util.List<com.capitalone.dashboard.model.Owner> cannot be converted to com.capitalone.dashboard.model.Owner
[ERROR] /workspace/hygieia-poojith/Hygieia/api/src/main/java/com/capitalone/dashboard/service/DashboardServiceImpl.java:[551,72] incompatible types: java.util.List<com.capitalone.dashboard.model.ActiveWidget> cannot be converted to java.util.List<java.lang.String>
[ERROR] /workspace/hygieia-poojith/Hygieia/api/src/main/java/com/capitalone/dashboard/service/DashboardServiceImpl.java:[553,111] incompatible types: java.util.List<com.capitalone.dashboard.model.ActiveWidget> cannot be converted to java.util.List<java.lang.String>
[ERROR] /workspace/hygieia-poojith/Hygieia/api/src/main/java/com/capitalone/dashboard/service/DashboardRemoteServiceImpl.java:[273,112] incompatible types: java.util.List<com.capitalone.dashboard.model.Owner> cannot be converted to com.capitalone.dashboard.model.Owner
[INFO] 4 errors
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] com.capitalone.dashboard:Hygieia 3.01.2-SNAPSHOT ... SUCCESS [  1.209 s]
[INFO] com.capitalone.dashboard:api 3.01.2-SNAPSHOT ....... FAILURE [ 14.132 s]
[INFO] com.capitalone.dashboard:api-audit 3.0.2-SNAPSHOT .. SKIPPED
[INFO] com.capitalone.dashboard:nfrr-audit-collector 3.0.2-SNAPSHOT SKIPPED
[INFO] com.capitalone.dashboard:rally-collector 3.0.2-SNAPSHOT SKIPPED
[INFO] com.capitalone.dashboard:artifactory-artifact-collector 3.0.2-SNAPSHOT SKIPPED
[INFO] com.capitalone.dashboard:bamboo-build-collector 3.0.2-SNAPSHOT SKIPPED
[INFO] com.capitalone.dashboard:jenkins-build-collector 3.0.2-SNAPSHOT SKIPPED
[INFO] com.capitalone.dashboard:jenkins-cucumber-test-collector 3.0.2-SNAPSHOT SKIPPED
[INFO] com.capitalone.dashboard:jenkins-codequality 3.0.2-SNAPSHOT SKIPPED
[INFO] com.capitalone.dashboard:sonar-codequality-collector 3.0.2-SNAPSHOT SKIPPED
[INFO] com.capitalone.dashboard:aws-cloud-collector 3.0.2-SNAPSHOT SKIPPED
[INFO] com.capitalone.dashboard:udeploy-deployment-collector 3.0.2-SNAPSHOT SKIPPED
[INFO] com.capitalone.dashboard:xldeploy-deployment-collector 3.0.2-SNAPSHOT SKIPPED
[INFO] com.capitalone.dashboard:jira-feature-collector 3.0.2-SNAPSHOT SKIPPED
[INFO] com.capitalone.dashboard:versionone-feature-collector 3.0.2-SNAPSHOT SKIPPED
[INFO] com.capitalone.dashboard:gitlab-feature-collector 3.0.2-SNAPSHOT SKIPPED
[INFO] com.capitalone.dashboard:chat-ops-collector 3.0.2-SNAPSHOT SKIPPED
[INFO] com.capitalone.dashboard:appdynamics-performance-collector 3.0.2-SNAPSHOT SKIPPED
[INFO] com.capitalone.dashboard:bitbucket-scm-collector 3.0.2-SNAPSHOT SKIPPED
[INFO] com.capitalone.dashboard:github-scm-collector 3.0.2-SNAPSHOT SKIPPED
[INFO] com.capitalone.dashboard:github-graphql-scm-collector 3.0.2-SNAPSHOT SKIPPED
[INFO] com.capitalone.dashboard:subversion-collector 3.0.2-SNAPSHOT SKIPPED
[INFO] com.capitalone.dashboard:gitlab-scm-collector 3.0.2-SNAPSHOT SKIPPED
[INFO] com.capitalone.dashboard:hpsm-cmdb-collector 3.0.2-SNAPSHOT SKIPPED
[INFO] com.capitalone.dashboard:nexus-iq-collector 3.0.2-SNAPSHOT SKIPPED
[INFO] com.capitalone.dashboard:score-collector 3.0.2-SNAPSHOT SKIPPED
[INFO] com.capitalone.dashboard:jira-xray-testresult-collector 3.0.2-SNAPSHOT SKIPPED
[INFO] Hygieia Publisher Plugin 2.1.9-SNAPSHOT ............ SKIPPED
[INFO] com.capitalone.dashboard:UI 3.0.2-SNAPSHOT ......... SKIPPED
[INFO] com.capitalone.dashboard:ui-tests 3.0.2-SNAPSHOT ... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  23.563 s
[INFO] Finished at: 2019-04-10T23:59:24-07:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.5.1:compile (default-compile) on project api: Compilation failure: Compilation failure:
[ERROR] /workspace/hygieia-poojith/Hygieia/api/src/main/java/com/capitalone/dashboard/request/DashboardRequest.java:[141,11] incompatible types: java.util.List<com.capitalone.dashboard.model.Owner> cannot be converted to com.capitalone.dashboard.model.Owner
[ERROR] /workspace/hygieia-poojith/Hygieia/api/src/main/java/com/capitalone/dashboard/service/DashboardServiceImpl.java:[551,72] incompatible types: java.util.List<com.capitalone.dashboard.model.ActiveWidget> cannot be converted to java.util.List<java.lang.String>
[ERROR] /workspace/hygieia-poojith/Hygieia/api/src/main/java/com/capitalone/dashboard/service/DashboardServiceImpl.java:[553,111] incompatible types: java.util.List<com.capitalone.dashboard.model.ActiveWidget> cannot be converted to java.util.List<java.lang.String>
[ERROR] /workspace/hygieia-poojith/Hygieia/api/src/main/java/com/capitalone/dashboard/service/DashboardRemoteServiceImpl.java:[273,112] incompatible types: java.util.List<com.capitalone.dashboard.model.Owner> cannot be converted to com.capitalone.dashboard.model.Owner
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :api
[poojith@poojith-hygeiya Hygieia]$

**Comment by stevegal** [comment]
@poojithkumar looks like you haven't installed the core properly. To do that checkout my branch of the core that is referenced in this PR and do a `mvn clean install`. This puts the core into the local repository and your build of the main Hygiene project should be able to be completed in the normal way.

**Comment by OneMoreByte** [comment]
@stevegal Where are you at with these two PRs?

**Comment by stevegal** [comment]
Well, what I have in place basically works (or did a while ago). I’m not sure that the main team want to follow the approach taken (I’m sure @rvema will provide confirm), so it leaves this pr in limbo atm.
I think it has highlighted that there is demand for this feature though.

**Comment by stevegal** [comment]
@rvema let me know when the refactor to multiple components is complete and I'll split up the changes between the various repositories.

**Comment by rvema** [comment]
@stevegal almost done, just 2 components need to be migrated. targetting for this weekend.

**Comment by HTAfonso** [comment]
@rvema @stevegal  , we have forked from master branch and build but couldn't get this feature work. Apparently it hasn't been merged? Do you have any plan to merge it near future?

**Comment by rvema** [comment]
@HTAfonso Its work in progress, considering the complexity of this change it is taking some time for both development and testing. This issue is closed as it has been recreated against the projects api and hygieia-core.

**Comment by rvema** [comment]
@HTAfonso  - Here are the PRs related to the same - [hygieia-core](https://github.com/Hygieia/hygieia-core/pull/40) and [api](https://github.com/Hygieia/api/pull/32)

**Comment by stevegal** [comment]
Just to reiterate... All I need (I think) is some help testing it - the more widgets I can get feedback on the better. The ui also requires changes (https://github.com/Hygieia/Hygieia/pull/3142) build all 3 together (mvn install)... build out your docker containers and spin up. Please let me know on any of the PR's (although the UI one is probably easier) of any issues and I'll fix them. I just need some help from the community to get these changed in!

**Comment by HTAfonso** [comment]
@stevegal the first findings through the UI test but please remember that I am not that much familiar with the UI so some finding I mentioned maybe expected behavior. 

1. I believe Quality widget name is changed as Code analyze and 3 new widgets appear Log, Rally and Team 
2. After creating team dashboard if you go to Add/Delete Widgets there is no available operation. Page comes with list of available widgets on left side but the previously chosen widgets don't appear on right side. Even if you click the available widgets nothing is added.
3. Build widget scroll down does not work only a limited number of available jobs listed. Butserach works fine.
4. The chosen job name is not coming as build name (It stays as build 1...2 whatever) and it is not linked to Jenkins like before.
5. As soon as you set up one of the built widget all widgets start to show the same data

**Comment by stevegal** [comment]
@HTAfonso thanks so far, and for re-igniting my fire to get this PR back into shape (this is a home time project for me, so it's quite difficult to get lots of data into my mongo db). It was all working before changes were made to the structure and I must have lost something in the rework. I say all working but...

1) think this is expected.
2) Yeah, then edit stuff changed dramatically from when I first raised the pr to now, must have not corrected properly. Thanks!
3) Think this is existing behaviour - we should fix outside of this PR I think
4) This was the easiest thing to get it working - I'll rework naming once we have something that works. I've just called then <widgetName><index> for now with the intention of having these user changeable in the future. Again, probably something for a follow on PR.
5) Definitely a bug! I'll have a debug through and see if I can spot the issue.

**Comment by pookimon** [comment]
One more issue
Feature widget don't show correct numbers for Open and WIP issues on Kanban board
